src/Microsoft.Sbom.Api/PackageDetails/ComponentDetailsUtils/RubyGemsUtils.cs

```diff
@@ -0,0 +1,291 @@
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text.RegularExpressions;
+using Microsoft.ComponentDetection.Contracts.BcdeModels;
+using Microsoft.Sbom.Api.Exceptions;
+using Microsoft.Sbom.Api.Output.Telemetry;
+using Microsoft.Sbom.Common;
+using Serilog;
+using File = System.IO.File;
+
+namespace Microsoft.Sbom.Api.PackageDetails;
+
+/// <summary>
+/// Utilities for retrieving information from rubygems packages that may not be present on the buildDropPath.
+/// </summary>
+public class RubyGemsUtils : IPackageManagerUtils<RubyGemsUtils>
+{
+    private const int ExecutableIndex = 1;
+
+    private readonly IFileSystemUtils fileSystemUtils;
+    private readonly ILogger log;
+    private readonly IRecorder recorder;
+
+    private readonly string rubyGemsPath;
+    private string[] potentialGemEnvPaths;
+
+    public RubyGemsUtils(IFileSystemUtils fileSystemUtils, ILogger log, IRecorder recorder)
+    {
+        this.fileSystemUtils = fileSystemUtils ?? throw new ArgumentNullException(nameof(fileSystemUtils));
+        this.log = log ?? throw new ArgumentNullException(nameof(log));
+        this.recorder = recorder ?? throw new ArgumentNullException(nameof(recorder));
+
+        this.rubyGemsPath = GetRubyGemsSpecificationsPath();
+    }
+
+    // Takes in a scanned component and attempts to find the associated gemspec file. If it is not found then it returns null.
+    public string? GetMetadataLocation(ScannedComponent scannedComponent)
+    {
+        var gemspecLocation = rubyGemsPath;
+
+        if (string.IsNullOrEmpty(gemspecLocation))
+        {
+            return null;
+        }
+
+        var componentName = scannedComponent.Component.PackageUrl?.Name.ToLower();
+        var componentVersion = scannedComponent.Component.PackageUrl?.Version;
+
+        var fullGemspecPath = Path.Join(gemspecLocation, $"{componentName}-{componentVersion}.gemspec");
+
+        if (fileSystemUtils.FileExists(fullGemspecPath))
+        {
+            return fullGemspecPath;
+        }
+        else
+        {
+            // Enumerate directories in the gemspec location and then search for the gemspec file in each directory
+            var directories = fileSystemUtils.GetDirectories(gemspecLocation);
+
+            foreach (var directory in directories)
+            {
+                fullGemspecPath = Path.Join(directory, $"{componentName}-{componentVersion}.gemspec");
+
+                if (fileSystemUtils.FileExists(fullGemspecPath))
+                {
+                    return Path.GetFullPath(fullGemspecPath);
+                }
+            }
+        }
+
+        return null;
+    }
+
+    public ParsedPackageInformation ParseMetadata(string gemspecLocation)
+    {
+        var supplierField = string.Empty;
+        var licenseField = string.Empty;
+
+        // Regex patterns for finding the relevant properties of the .gemspec file.
+        var namePattern = @"s.name\s*=\s*""([^""]+)""";
```

Comment:
Are these single-lines delimited by newline? If so I'd suggest something like `fileContent.Split(Newline).Where(l => l.StartsWith("s.name")).Split('=')` (of course you'd need to harden it a bit). When possible I suggest avoiding Regex because it has a lot of caveats and security issues (RegexDoS) that make it hard to maintain and debug.

Even better would be if there was a C# library which already parses GemSpec for us.

Reply:
I couldn't find one in C# but there is one in Ruby. We could use something like:

using IronRuby;
using IronRuby.Runtime;
using Microsoft.Scripting.Hosting;

// Create a Ruby engine
var engine = Ruby.CreateEngine();

// Execute the .gemspec file and get the Gem::Specification object
var spec = engine.ExecuteFile("my_gem.gemspec") as RubyObject;

but how much overhead this Ruby engine would add.

Reply:
Their last release was 2011 so I wouldn't want to take a dependency on them. Either the Regex or a split-and-search should be fine. If you stick with Regex I'd suggest adding a timeout to the Regex.Match call just as a precaution. These patterns don't look vulnerable to catastrophic backtracking, but I'm not a Regex wizard so I'm not 100% sure.